File: aetherctl/src/openspec/control.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import { open, unlink } from "node:fs/promises";
import { computePlanDigest, ensureState, loadState, newRunId, updateState } from "./state.js";
import { runChange } from "./runner.js";
import { safePath, sanitizePathComponent } from "./security.js";

const pExecFile = promisify(execFile);

/**
 * Run manager to handle concurrent execution across multiple changes.
 * Uses file-based locks instead of in-memory state for robustness.
 */
class RunManager {
  private runningLocks = new Map<string, { lockPath: string; release: () => Promise<void> }>();

  async tryAcquireRun(repoPath: string, changeId: string): Promise<(() => Promise<void>) | null> {
    const key = `${repoPath}:${changeId}`;

    // Check if already running in this process
    if (this.runningLocks.has(key)) {
      return null;
    }

    try {
      const sanitizedChangeId = sanitizePathComponent(changeId);
      const lockPath = safePath(repoPath, "openspec", "changes", sanitizedChangeId, "run.lock");

      // Try to acquire lock (non-blocking)
      const fh = await open(lockPath, "wx");
      await fh.writeFile(`${process.pid}\n${Date.now()}\n`, "utf8");
      await fh.close();

      const release = async () => {
        try {
          await unlink(lockPath);
          this.runningLocks.delete(key);
        } catch (e) {
          console.error(`[openspec] Failed to release run lock: ${e}`);
        }
      };

      this.runningLocks.set(key, { lockPath, release });
      return release;
    } catch (e: any) {
      if (e?.code === "EEXIST") {
        return null; // Already running
      }
      throw e;
    }
  }
}

const runManager = new RunManager();

export async function approveStart(opts: {
  repoPath: string;
  changeId: string;
  requestedBy?: string | null;
}): Promise<{ ok: true; status: string } | { ok: false; error: string }> {
  const repoPath = opts.repoPath;
  const changeId = opts.changeId;

  const st = await ensureState(repoPath, changeId);
  if (st.status !== "WAIT_APPROVAL") {
    return { ok: false, error: `Change not in WAIT_APPROVAL (current: ${st.status})` };
  }

  // Try to acquire run lock
  const release = await runManager.tryAcquireRun(repoPath, changeId);
  if (!release) {
    return { ok: false, error: `BUSY (another process is running this change)` };
  }

  try {
    const digest = await computePlanDigest(repoPath, changeId);
    const headSha = await gitHeadSha(repoPath);
    const runId = newRunId();

    await updateState(repoPath, changeId, (prev) => {
      const now = new Date().toISOString();
      return {
        ...prev,
        plan: { planDigest: digest, headSha },
        approval: {
          approved: true,
          approvedAt: now,
          approvedBy: opts.requestedBy ?? null,
          approvedPlanDigest: digest
        },
        blocked: {
          ...prev.blocked,
          reason: null,
          needs: null,
          pause: {
            kind: null,
            request: null,
            unblockCommand: null,
            resumeCondition: null,
            userInput: null,
            requestedAt: null,
            resolvedAt: null
          }
        },
        execution: {
          ...prev.execution,
          runId,
          attempt: (prev.execution.attempt ?? 0) + 1,
          startedAt: now,
          lastHeartbeatAt: now
        },
        status: "PRECHECK"
      };
    });

    const res = await runChange(repoPath, changeId);
    return res.ok ? { ok: true, status: res.status } : { ok: false, error: res.error };
  } finally {
    await release();
  }
}

export async function resumeChange(opts: {
  repoPath: string;
  changeId: string;
  userInput: string;
  requestedBy?: string | null;
}): Promise<{ ok: true; status: string } | { ok: false; error: string }> {
  const repoPath = opts.repoPath;
  const changeId = opts.changeId;

  const st = await ensureState(repoPath, changeId);
  if (["DONE", "CANCELLED"].includes(st.status)) return { ok: true, status: st.status };
  if (st.status === "FAILED") return { ok: false, error: "Change already FAILED" };
  if (st.status !== "PAUSED_UNEXPECTED") return { ok: false, error: `Change not in PAUSED_UNEXPECTED (current: ${st.status})` };
  if (!st.blocked.pause.request) return { ok: false, error: "No pause request present" };

  // Try to acquire run lock
  const release = await runManager.tryAcquireRun(repoPath, changeId);
  if (!release) {
    return { ok: false, error: `BUSY (another process is running this change)` };
  }

  try {
    const runId = st.execution.runId ?? newRunId();
    await updateState(repoPath, changeId, (prev) => {
      const now = new Date().toISOString();
      return {
        ...prev,
        blocked: {
          ...prev.blocked,
          pause: {
            ...prev.blocked.pause,
            userInput: opts.userInput,
            resolvedAt: now
          }
        },
        error: { code: null, message: null, detail: null, retryable: null, firstSeenAt: prev.error.firstSeenAt, lastSeenAt: null },
        execution: { ...prev.execution, runId, lastHeartbeatAt: now },
        status: "PRECHECK"
      };
    });

    const res = await runChange(repoPath, changeId);
    return res.ok ? { ok: true, status: res.status } : { ok: false, error: res.error };
  } finally {
    await release();
  }
}

export async function cancelChange(opts: {
  repoPath: string;
  changeId: string;
}): Promise<{ ok: true; status: string } | { ok: false; error: string }> {
  const st = await ensureState(opts.repoPath, opts.changeId);
  if (["DONE", "CANCELLED", "FAILED"].includes(st.status)) return { ok: true, status: st.status };

  const next = await updateState(opts.repoPath, opts.changeId, (prev) => {
    if (["WAIT_APPROVAL", "PLANNING", "PAUSED_UNEXPECTED"].includes(prev.status)) {
      return { ...prev, status: "CANCELLED" };
    }
    return { ...prev, status: "CANCEL_REQUESTED" };
  });

  return { ok: true, status: next.status };
}

export async function statusChange(opts: {
  repoPath: string;
  changeId: string;
}): Promise<{ ok: true; state: unknown; progressTail?: string[] } | { ok: false; error: string }> {
  try {
    const st = await loadState(opts.repoPath, opts.changeId);
    return { ok: true, state: st };
  } catch (e: any) {
    return { ok: false, error: e?.message ?? String(e) };
  }
}

async function gitHeadSha(repoPath: string): Promise<string> {
  const { stdout } = await pExecFile("git", ["rev-parse", "HEAD"], { cwd: repoPath });
  return stdout.trim();
}

File: nanobot/agent/taskmode/control.py
"""Control commands: /approve, /cancel, /status, /resume."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from loguru import logger

from nanobot.agent.taskmode.state import (
    CANCELLED,
    CANCEL_REQUESTED,
    DONE,
    EXECUTING,
    FAILED,
    PAUSED_UNEXPECTED,
    PLANNING,
    PRECHECK,
    RUNNING_STATUSES,
    TERMINAL_STATUSES,
    VERIFYING,
    WAIT_APPROVAL,
    StateManager,
    new_run_id,
)


@dataclass
class ControlCommand:
    action: str                 # approve | cancel | status | resume
    change_id: str | None = None
    user_input: str | None = None
    requested_by: str | None = None


# ---------------------------------------------------------------------------
# Parsing
# ---------------------------------------------------------------------------

def parse_control_command(text: str) -> ControlCommand | None:
    """Parse a control slash command.  Returns ``None`` if not a control cmd."""
    text = text.strip()

    if text.startswith("/approve"):
        parts = text.split(maxsplit=1)
        cid = parts[1].strip() if len(parts) > 1 else None
        return ControlCommand(action="approve", change_id=cid)

    if text.startswith("/cancel"):
        parts = text.split(maxsplit=1)
        cid = parts[1].strip() if len(parts) > 1 else None
        return ControlCommand(action="cancel", change_id=cid)

    if text.startswith("/status"):
        parts = text.split(maxsplit=1)
        cid = parts[1].strip() if len(parts) > 1 else None
        return ControlCommand(action="status", change_id=cid)

    if text.startswith("/resume"):
        parts = text.split(maxsplit=2)
        cid = parts[1].strip() if len(parts) > 1 else None
        user_input = parts[2].strip() if len(parts) > 2 else ""
        return ControlCommand(action="resume", change_id=cid, user_input=user_input)

    return None


# ---------------------------------------------------------------------------
# Execution
# ---------------------------------------------------------------------------

async def execute_control(
    cmd: ControlCommand,
    workspace: Path,
    runner_factory: Any = None,
) -> str:
    """Dispatch a control command and return a human-readable response."""
    state_mgr = StateManager()

    if cmd.action == "approve":
        return await _handle_approve(cmd, workspace, state_mgr, runner_factory)
    if cmd.action == "cancel":
        return _handle_cancel(cmd, workspace, state_mgr)
    if cmd.action == "status":
        return _handle_status(cmd, workspace, state_mgr)
    if cmd.action == "resume":
        return await _handle_resume(cmd, workspace, state_mgr, runner_factory)

    return f"Unknown control command: {cmd.action}"


# ---------------------------------------------------------------------------
# Handlers
# ---------------------------------------------------------------------------

def _resolve_change_id(cmd: ControlCommand, workspace: Path, state_mgr: StateManager) -> str | None:
    """Resolve change_id from command or auto-detect the single active change."""
    if cmd.change_id:
        return cmd.change_id

    active = state_mgr.list_active(workspace)
    if len(active) == 1:
        return active[0].change_id
    if len(active) == 0:
        return None
    return None  # Ambiguous


async def _handle_approve(
    cmd: ControlCommand,
    workspace: Path,
    state_mgr: StateManager,
    runner_factory: Any,
) -> str:
    change_id = _resolve_change_id(cmd, workspace, state_mgr)
    if not change_id:
        active = state_mgr.list_active(workspace)
        if not active:
            return "No active changes to approve."
        ids = ", ".join(s.change_id for s in active)
        return f"Multiple active changes. Specify one: {ids}"

    change_dir = state_mgr.change_dir(workspace, change_id)
    if not state_mgr.state_path(change_dir).exists():
        return f"Change `{change_id}` not found."

    state = state_mgr.read(change_dir)
    if state.status != WAIT_APPROVAL:
        return f"Cannot approve: current status is **{state.status}**."

    # Prevent concurrent runs
    running = [s for s in state_mgr.list_active(workspace) if s.status in RUNNING_STATUSES]
    if running:
        return f"Cannot approve: change `{running[0].change_id}` is currently running."

    # Set approval
    from datetime import datetime
    digest = state_mgr.compute_plan_digest(change_dir)
    now = datetime.utcnow().isoformat() + "Z"

    state.approval.approved = True
    state.approval.approved_at = now
    state.approval.approved_by = cmd.requested_by
    state.approval.approved_plan_digest = digest
    state.plan.plan_digest = digest
    state.plan.head_sha = _git_head_sha(workspace)
    state.execution.run_id = new_run_id()
    state.execution.attempt += 1
    state.execution.started_at = now
    state.execution.last_heartbeat_at = now
    state.status = PRECHECK
    # Clear blocked
    from nanobot.agent.taskmode.state import BlockedInfo, PauseInfo
    state.blocked = BlockedInfo()
    state_mgr.write(change_dir, state)

    # Run the task
    if runner_factory:
        runner = runner_factory()
        result = await runner.run(change_id)
        return result

    return f"Change `{change_id}` approved and transitioned to PRECHECK."


def _handle_cancel(
    cmd: ControlCommand,
    workspace: Path,
    state_mgr: StateManager,
) -> str:
    change_id = _resolve_change_id(cmd, workspace, state_mgr)
    if not change_id:
        active = state_mgr.list_active(workspace)
        if not active:
            return "No active changes to cancel."
        ids = ", ".join(s.change_id for s in active)
        return f"Multiple active changes. Specify one: {ids}"

    change_dir = state_mgr.change_dir(workspace, change_id)
    if not state_mgr.state_path(change_dir).exists():
        return f"Change `{change_id}` not found."

    state = state_mgr.read(change_dir)

    if state.status in TERMINAL_STATUSES:
        return f"Change `{change_id}` is already **{state.status}**."

    # Immediate cancel for non-running states
    if state.status in {WAIT_APPROVAL, PLANNING, PAUSED_UNEXPECTED}:
        state.status = CANCELLED
        state_mgr.write(change_dir, state)
        return f"Change `{change_id}` cancelled."

    # For running states, request cancellation (loop will pick it up)
    state.status = CANCEL_REQUESTED
    state_mgr.write(change_dir, state)
    return f"Cancellation requested for `{change_id}` (currently {state.status})."


def _handle_status(
    cmd: ControlCommand,
    workspace: Path,
    state_mgr: StateManager,
) -> str:
    # If a specific change is requested
    if cmd.change_id:
        change_dir = state_mgr.change_dir(workspace, cmd.change_id)
        if not state_mgr.state_path(change_dir).exists():
            return f"Change `{cmd.change_id}` not found."
        state = state_mgr.read(change_dir)
        return _format_status(state)

    # Otherwise list all active changes
    active = state_mgr.list_active(workspace)
    if not active:
        return "No active changes."

    parts = []
    for s in active:
        parts.append(_format_status(s))
    return "\n\n---\n\n".join(parts)


async def _handle_resume(
    cmd: ControlCommand,
    workspace: Path,
    state_mgr: StateManager,
    runner_factory: Any,
) -> str:
    change_id = _resolve_change_id(cmd, workspace, state_mgr)
    if not change_id:
        active = state_mgr.list_active(workspace)
        if not active:
            return "No active changes to resume."
        ids = ", ".join(s.change_id for s in active)
        return f"Multiple active changes. Specify one: {ids}"

    change_dir = state_mgr.change_dir(workspace, change_id)
    if not state_mgr.state_path(change_dir).exists():
        return f"Change `{change_id}` not found."

    state = state_mgr.read(change_dir)

    if state.status in {DONE, CANCELLED}:
        return f"Change `{change_id}` is already **{state.status}**."
    if state.status == FAILED:
        return f"Change `{change_id}` has **FAILED** and cannot be resumed."
    if state.status != PAUSED_UNEXPECTED:
        return f"Cannot resume: current status is **{state.status}** (expected PAUSED_UNEXPECTED)."

    # Resolve pause
    from datetime import datetime
    now = datetime.utcnow().isoformat() + "Z"

    state.blocked.pause.user_input = cmd.user_input
    state.blocked.pause.resolved_at = now
    state.error.code = None
    state.error.message = None
    state.error.detail = None
    state.error.retryable = None
    state.error.last_seen_at = None
    state.execution.last_heartbeat_at = now
    state.status = PRECHECK
    state_mgr.write(change_dir, state)

    if runner_factory:
        runner = runner_factory()
        result = await runner.run(change_id)
        return result

    return f"Change `{change_id}` resumed and transitioned to PRECHECK."


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _format_status(state: Any) -> str:
    """Format a TaskState into a readable status string."""
    lines = [
        f"**Change**: `{state.change_id}`",
        f"**Status**: {state.status}",
    ]
    if state.execution.current_task:
        lines.append(f"**Current task**: {state.execution.current_task}")
    if state.loop.ticks:
        lines.append(f"**Progress**: tick {state.loop.ticks}/{state.loop.max_ticks}")
    if state.error.message:
        lines.append(f"**Error**: {state.error.message}")
    if state.blocked.reason:
        lines.append(f"**Blocked**: {state.blocked.reason}")
    return "\n".join(lines)


def _git_head_sha(workspace: Path) -> str | None:
    """Get the current git HEAD SHA, or None."""
    import subprocess
    try:
        result = subprocess.run(
            ["git", "rev-parse", "HEAD"],
            cwd=str(workspace),
            capture_output=True,
            text=True,
            timeout=10,
        )
        if result.returncode == 0:
            return result.stdout.strip()
    except Exception:
        pass
    return None
